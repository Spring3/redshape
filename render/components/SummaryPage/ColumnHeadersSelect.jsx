import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import Select from 'react-select';
import { connect } from 'react-redux';
import makeAnimated from 'react-select/lib/animated';
import { withTheme } from 'styled-components';

import actions from '../../actions';
import { Label } from '../Input';

const selectStyles = {
  container: (base, state) => {
    return { ...base };
  },
  multiValue: (base, state) => {
    return state.data.isFixed
      ? { ...base, backgroundColor: '#FAFAFA', border: '1px solid #A4A4A4' }
      : { ...base, backgroundColor: 'transparent', border: '1px solid #3F3844' };
  },
  multiValueLabel: (base, state) => {
    return state.data.isFixed
      ? { ...base, paddingRight: 6, color: '#A4A4A4' }
      : base;
  },
  multiValueRemove: (base, state) => {
    return state.data.isFixed ? { ...base, display: 'none' } : base;
  }
};

class ColumnHeadersSelect extends Component {
  constructor(props) {
    super(props);

    this.issueHeaders = [
      { label: 'Id', isFixed: true, value: 'id' },
      { label: 'Project', value: 'project.name' },
      { label: 'Tracker', value: 'tracker.name', format: 'tracker' },
      { label: 'Status', value: 'status.name', format: 'status' },
      { label: 'Subject', isFixed: true, value: 'subject' },
      { label: 'Priority', value: 'priority.name', format: 'priority' },
      { label: 'Estimation', value: 'estimated_hours', format: 'hours' },
      { label: 'Total Estimation', value: 'total_estimated_hours', format: 'hours' },
      { label: 'Spent', value: 'spent_hours', format: 'hours' },
      { label: 'Total Spent', value: 'total_spent_hours', format: 'hours' },
      { label: 'Progress', value: 'done_ratio', format: 'progress' },
      { label: 'Due Date', value: 'due_date', format: 'date' }
    ];
  }

  onHeadersSelectChange = (value, { action, removedValue }) => {
    switch (action) {
      case 'remove-value':
      case 'pop-value':
        if (removedValue.isFixed) {
          return;
        }
        break;
    }

    this.props.settingsChangeIssueHeaders(value);
  }

  render() {
    const { theme, issueHeaders } = this.props;
    return (
      <Label htmlFor="headers" label="Table Columns">
        <Select
          name="headers"
          styles={selectStyles}
          components={makeAnimated()}
          options={this.issueHeaders}
          defaultValue={issueHeaders}
          value={issueHeaders}
          onChange={this.onHeadersSelectChange}
          isMulti={true}
          isClearable={false}
          theme={(defaultTheme) => ({
            ...defaultTheme,
            borderRadius: 3,
            colors: {
              ...defaultTheme.colors,
              primary: theme.main,
            },
          })
          }
        />
      </Label>
    );
  }
}

ColumnHeadersSelect.propTypes = {
  issueHeaders: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    format: PropTypes.string,
    isFixed: PropTypes.bool
  }).isRequired).isRequired,
  settingsChangeIssueHeaders: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  issueHeaders: state.settings.issueHeaders,
});

const mapDispatchToProps = dispatch => ({
  settingsChangeIssueHeaders: issueHeaders => dispatch(actions.settings.setIssueHeaders(issueHeaders))
});

export default withTheme(connect(mapStateToProps, mapDispatchToProps)(ColumnHeadersSelect));
