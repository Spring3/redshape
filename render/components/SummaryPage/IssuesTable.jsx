import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import _get from 'lodash/get';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import styled, { withTheme, css } from 'styled-components';
import SortAscendingIcon from 'mdi-react/SortAscendingIcon';
import SortDescendingIcon from 'mdi-react/SortDescendingIcon';

import InfiniteScroll from '../InfiniteScroll';
import ProcessIndicator, { OverlayProcessIndicator } from '../ProcessIndicator';
import Date from '../Date';
import Progressbar from "../Progressbar";

import { Tag, Status, IssueId, Priority } from "../Issue";

const Table = styled.table`
  position: relative;
  min-height: 200px;
  width: 100%;
  border: 2px solid ${props => props.theme.bgDark};
  border-spacing: 0px;

  tbody {
    text-align: center;
    font-size: 14px;
    overflow-y: scroll;
    max-height: 60vh;

    tr {
      td {
        border-bottom: 2px solid ${props => props.theme.bgDark};
        padding: 15px 5px;
        transition: background ease ${props => props.theme.transitionTime};
      }

      &:hover {
        cursor: pointer;
        
        td {
          background: ${props => props.theme.bgDark};
        }
      }

      td.due_date {
        white-space: nowrap;
      }
      
      td.subject {
        text-align: left;
      }
    }
  }

  thead {
    z-index: 1;
    tr {
      th {
        padding: 15px 5px;
        background: ${props => props.theme.bgDark};
        border-bottom: 2px solid ${props => props.theme.bgDarker};
        transition: color ease ${props => props.theme.transitionTime};
        &:hover {
          cursor: pointer;
          color: ${props => props.theme.main};
        }
  
        svg {
          vertical-align: middle;
        }
        
        &.subject {
          text-align: left;
        }
      }
    }
  }
`;

const ProcessIndicatorContainer = styled.tr`
  position: absolute;
  width: 100%;
  text-align: center;

  div.container {
    position: absolute;
    top: 15px;
    left: 45%;
  }
`;

const ColorfulSpan = styled.span`
  ${
    props => props.color
    ? css `
      padding-bottom: 2px;
      background: linear-gradient(to bottom,transparent 0,transparent 90%,${props.color} 90%,${props.color} 100%);
    `
    : null
  }
`;

const colorMap = {
  'closed': 'red',
  'high': 'yellow',
  'urgent': 'red',
  'immediate': 'red',
  'critical': 'red',
  'open': 'green',
  'low': 'green',
  'pending': 'yellow',
  'normal': 'yellow'
};

class IssuesTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sortBy: undefined,
      sortDirection: undefined
    };
  }

  sortTable = (by) => () => {
    const { onSort } = this.props;
    const { sortBy, sortDirection } = this.state;
    if (sortBy !== by) {
      this.setState({
        sortBy: by,
        sortDirection: 'asc'
      }, () => onSort(by, 'asc'));
    } else {
      const newSortDirection = sortDirection === 'asc' ? 'desc' : 'asc';
      this.setState({ sortDirection: newSortDirection }, () => onSort(by, newSortDirection));
    }
  }

  showIssueDetails(id) {
    this.props.history.push(`/app/issue/${id}/`);
  }

  paint = (format, value, textValue, opts) => {
    const { theme, uiStyle } = this.props;
    const isEnhanced = opts.isEnhanced;

    if (isEnhanced){
      if (format === 'id' && isEnhanced) {
        return (<IssueId value={value} tracker={opts.tracker}/>);
      }else if (format === 'status' && isEnhanced) {
        return (<Status simple={true} value={value}/>);
      }else if (format === 'progress' && isEnhanced) {
        return (<Progressbar percent={value} width={80} mode="progress-gradient" height={7}/>);
      }else if (format === 'priority' && isEnhanced) {
        return (<Priority centered value={value}/>);
      }
    }

    if (format === 'date'){
      return (<Date date={value}/>);
    }else if (format === 'tags' && value){
      return (<Fragment>{ value ? value.map(el => <Tag key={el} mode={isEnhanced ? "simple" : "plain"} value={el}></Tag>) : undefined}</Fragment>);
    }else{
      const color = (uiStyle === 'colors' && typeof value === 'string'
        ? colorMap[value.toLowerCase()]
        : undefined);

      return (
        <ColorfulSpan color={theme[color]} className={opts && opts.className}>{textValue}</ColorfulSpan>
      );
    }
  }

  loadIssuePage = () => {
    const { issues, fetchIssuePage } = this.props;
    const { page } = issues;
    fetchIssuePage(page + 1);
  }

  render() {
    const { issueHeaders, issues, limit, uiStyle } = this.props;
    const { sortBy, sortDirection } = this.state;
    let userTasks = issues.data;
    if (limit){
      userTasks = userTasks.slice(0,  limit)
    }
    const isEnhanced = uiStyle === 'enhanced';
    return (
      <Fragment>
        { (!userTasks.length && issues.isFetching) && (<OverlayProcessIndicator />) }
        <Table>
          <thead>
          <tr>
            {issueHeaders.map(header => (
              <th
                key={header.value}
                onClick={this.sortTable(header.value)}
                className={isEnhanced ? header.value : ""}
              >
                {header.label}&nbsp;
                {sortBy === header.value && sortDirection === 'asc' && (
                  <SortAscendingIcon size={14} />
                )}
                {sortBy === header.value && sortDirection === 'desc' && (
                  <SortDescendingIcon size={14} />
                )}
              </th>
            ))}
          </tr>
          </thead>
          <tbody>
          <InfiniteScroll
            load={this.loadIssuePage}
            isEnd={userTasks.length === issues.totalCount || limit === userTasks.length}
            hasMore={!issues.isFetching && !issues.error && userTasks.length < issues.totalCount && !limit}
            container={window}
            loadIndicator={<ProcessIndicatorContainer><td><ProcessIndicator className="container" /></td></ProcessIndicatorContainer>}
            immediate={true}
          >
            {userTasks.map((task)  => (
              <tr key={task.id} onClick={this.showIssueDetails.bind(this, task.id)}>
                {
                  issueHeaders.map(header => {
                    let opts = {};
                    if (isEnhanced) {
                      opts.isEnhanced = isEnhanced;
                      opts.tracker = _get(task, 'tracker.id');
                    }
                    const format = header.format;
                    let value = _get(task, header.value);
                    let textValue = value;
                    if (format === 'hours'){
                      if (value != null){
                        textValue = Number(value.toFixed(2));
                      }
                    }else if (format === 'progress'){
                      textValue = `${value}%`
                    }else if (format === 'count' && value && value.length){
                      textValue = `${value.length}`
                    }
                    return (
                      <td
                        key={header.value}
                        className={isEnhanced ? header.value : (header.value === 'due_date' ? header.value : "")}
                      >
                        { this.paint(format, value, textValue, opts) }
                      </td>
                    )})
                }
              </tr>
            ))}
          </InfiniteScroll>
          </tbody>
        </Table>
      </Fragment>
    );
  }
}

IssuesTable.propTypes = {
  issues: PropTypes.shape({
    userTasks: PropTypes.arrayOf(PropTypes.object).isRequired,
  }).isRequired,
  uiStyle: PropTypes.string.isRequired,
  issueHeaders: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    isFixed: PropTypes.bool
  }).isRequired).isRequired,
  onSort: PropTypes.func.isRequired,
  fetchIssuePage: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  uiStyle: state.settings.uiStyle,
  issues: state.issues.all,
  issueHeaders: state.settings.issueHeaders,
});

export default withRouter(withTheme(connect(mapStateToProps)(IssuesTable)));
