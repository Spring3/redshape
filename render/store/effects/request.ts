import { ipcRenderer, remote } from 'electron';
import { Response } from '../../../types';

const crypto = remote.require('crypto');

type Config = {
  endpoint: string;
  token?: string;
}

type QueryConfig = {
  payload: Record<string, any>;
  config?: Config;
}

const query = (config: QueryConfig): Promise<Response> => {
  const id = crypto.randomBytes(10).toString('hex');

  return new Promise((resolve) => {
    ipcRenderer.send('request', JSON.stringify({ ...config, id }));

    ipcRenderer.once(`response:${id}`, (event, response) => {
      console.log('response:', response);
      resolve(response);
    });
  });
};

export {
  query,
};

export type {
  QueryConfig
};
