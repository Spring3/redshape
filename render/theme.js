export default {
  bg: '#FFFFFF',
  bgDark: '#F9F9F9',
  bgDarker: '#E9E9E9',
  bgDisabled: 'hsl(0,0%,95%)',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#A0A0A0',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  shadow: '#D0D0D0',
  transitionTime: '.2s'
};
