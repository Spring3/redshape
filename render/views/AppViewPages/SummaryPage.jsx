import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import { connect } from 'react-redux';
import styled, { withTheme } from 'styled-components';
import MagnifyIcon from 'mdi-react/MagnifyIcon';

import { IssueFilter } from '../../actions/helper';
import actions from '../../actions';
import { Input } from '../../components/Input';
import IssuesTable from '../../components/SummaryPage/IssuesTable';

const Grid = styled.div`
  display: grid;
  padding: 20px;
  grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
  margin-bottom: 60px;
`;

const Section = styled.section`
  background: white;
  padding: 0px 20px 20px 20px;
  border-radius: 5px;
`;

const IssuesSection = styled(Section)`
  grid-column: span 8;
  grid-row: auto;
`;

const OptionsGrid = styled.div`
  display: grid;
  // grid-template-columns: minmax(200px, auto) 1fr;
  grid-template-columns: 1fr minmax(80px, auto);
  // grid-template-columns: auto auto;
  grid-template-rows: auto;
  grid-row-gap: 20px;
  grid-column-gap: 20px;
  margin-bottom: 20px;

  // background-color: #EFEFEF;
  // border-radius: 4px;
  // border: 1px solid #A0A0A0;
`;

const Box = styled.div`
  align-self: center;
`;

class SummaryPage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      search: undefined,
      sortBy: undefined,
      sortDirection: undefined,
      searchIncludeID: false
    };

    this.deboucedFetch = debounce(this.fetchIssues, 500);
  }

  componentWillMount() {
    this.fetchIssues();
  }

  componentDidUpdate(oldProps) {
    if (oldProps.showClosedIssues !== this.props.showClosedIssues) {
      this.fetchIssues();
    }
  }

  fetchIssues = (page = 0) => {
    const { search, sortBy, sortDirection } = this.state;
    const { userId, showClosedIssues } = this.props;
    if (userId) {
      const showByAuthor = this.props.mode === 'author';
      let queryFilter = new IssueFilter();
      if (showByAuthor) {
        queryFilter = queryFilter.author(userId);
      } else {
        queryFilter = queryFilter.assignee(userId);
      }
      queryFilter = queryFilter
        .status({ open: true, closed: showClosedIssues })
        .title(search)
        .sort(sortBy, sortDirection)
        .build();
      this.props.fetchIssues(queryFilter, page);
    }
  }

  onRefresh = () => {
    this.fetchIssues();
  }

  onSearchChange = (e) => {
    this.setState({
      search: e.target.value
    }, () => this.deboucedFetch());
  }

  onSort = (sortBy, sortDirection) => {
    this.setState({
      sortBy,
      sortDirection
    }, () => this.deboucedFetch());
  }

  toggleClosedIssuesDisplay = () => {
    const { settingsShowClosedIssues, showClosedIssues } = this.props;
    settingsShowClosedIssues(!showClosedIssues);
  }

  render() {
    const { theme, showClosedIssues, mode } = this.props;
    return (
      <Grid>
        <IssuesSection>
          <h2>
            { mode === 'author' ? 'Issues created by me' : 'Issues assigned to me'}
          </h2>
          <OptionsGrid>
            <Input
              icon={(
                <MagnifyIcon
                  xmlns="http://www.w3.org/2000/svg"
                  color={theme.main}
                />
)}
              type="text"
              name="search"
              placeholder="Search..."
              onChange={this.onSearchChange}
            />
            <Box>
              <label>
                <Input
                  type="checkbox"
                  checked={showClosedIssues}
                  onChange={this.toggleClosedIssuesDisplay}
                />
                <span>Closed</span>
              </label>
            </Box>
          </OptionsGrid>
          <IssuesTable
            onSort={this.onSort}
            fetchIssuePage={this.fetchIssues}
          />
        </IssuesSection>
      </Grid>
    );
  }
}

SummaryPage.propTypes = {
  userId: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  showClosedIssues: PropTypes.bool.isRequired,
  fetchIssues: PropTypes.func.isRequired,
  theme: PropTypes.object.isRequired,
  settingsShowClosedIssues: PropTypes.func.isRequired,
  mode: PropTypes.string,
};

const mapStateToProps = state => ({
  userId: state.user.id,
  showClosedIssues: state.settings.showClosedIssues
});

const mapDispatchToProps = dispatch => ({
  fetchIssues: (filter, page) => dispatch(actions.issues.getPage(filter, page)),
  settingsShowClosedIssues: value => dispatch(actions.settings.setShowClosedIssues(value)),
});

export default withTheme(connect(mapStateToProps, mapDispatchToProps, null, { forwardRef: true })(SummaryPage));
