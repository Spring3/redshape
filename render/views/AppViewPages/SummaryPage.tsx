import React, { useState } from 'react';
import styled, { useTheme } from 'styled-components';
import MagnifyIcon from 'mdi-react/MagnifyIcon';

import { Input } from '../../components/Input';
import { IssuesTable } from '../../components/SummaryPage/IssuesTable';
import { Flex } from '../../components/Flex';
import { useOvermindActions, useOvermindState } from '../../store';

const Grid = styled.div`
  display: grid;
  padding: 20px;
  grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));
  margin-bottom: 60px;
`;

const Section = styled.section`
  background: white;
  padding: 0px 20px 20px 20px;
  border-radius: 5px;
`;

const IssuesSection = styled(Section)`
  grid-column: span 8;
  grid-row: auto;
`;

const OptionsGrid = styled.div`
  display: grid;
  grid-template-columns: minmax(200px, auto) 1fr;
  grid-template-rows: auto;
  grid-row-gap: 20px;
  grid-column-gap: 20px;
  margin-bottom: 20px;
`;

const GridRow = styled.div`
  grid-column: 1/-1;
`;

const SummaryPage = () => {
  const [search, setSearch] = useState<string>('');

  const theme = useTheme();
  const state = useOvermindState();
  const actions = useOvermindActions();

  const onSearchChange = (e: any) => {
    setSearch(e.target.value);
  };

  const toggleClosedIssues = async () => {
    await actions.settings.update({
      ...state.settings,
      showClosedIssues: !state.settings.showClosedIssues
    });
  };

  return (
    <Grid>
      <IssuesSection>
        <h2>Issues assigned to me</h2>
        <OptionsGrid>
          <GridRow>
            <Flex alignItems="center">
              <Input
                icon={(
                  <MagnifyIcon
                    xmlns="http://www.w3.org/2000/svg"
                    color={(theme as any).main}
                  />
                  )}
                type="text"
                name="search"
                placeholder="Search..."
                onChange={onSearchChange}
              />
              <label>
                <Input
                  type="checkbox"
                  checked={state.settings.showClosedIssues}
                  onChange={toggleClosedIssues}
                />
                <span>Include closed</span>
              </label>
            </Flex>
          </GridRow>
        </OptionsGrid>
        <IssuesTable search={search} />
      </IssuesSection>
    </Grid>
  );
};

export {
  SummaryPage
};
