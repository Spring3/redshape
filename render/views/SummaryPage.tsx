import React, { useEffect, useState } from 'react';
import { css } from '@emotion/react';

import { debounce } from 'lodash';
import { Input } from '../components/Input';
import { IssuesTable } from '../components/SummaryPage/IssuesTable';
import { Flex } from '../components/Flex';
import { useOvermindActions, useOvermindState } from '../store';
import { useNavbar } from '../contexts/NavbarContext';
import { Checkbox } from '../components/Checkbox';

const SummaryPage = () => {
  const [search, setSearch] = useState<string>('');

  const state = useOvermindState();
  const actions = useOvermindActions();
  const navbarState = useNavbar();

  useEffect(() => {
    navbarState.setTitle('My Backlog');
  }, [navbarState.setTitle]);

  const onSearchChange = debounce((e: any) => {
    setSearch(e.target.value);
  }, 200);

  const toggleClosedIssues = async () => {
    await actions.settings.update({
      ...state.settings,
      showClosedIssues: !state.settings.showClosedIssues
    });
  };

  return (
    <Flex
      css={css`
        margin-bottom: 2rem;
        margin-top: 3.5rem;
        padding-left: 0.5rem;
        padding-right: 0.5rem;
      `}
      direction="column"
    >
      <Flex
        grow="100%"
        css={css`
          padding-left: 1rem;
          padding-right: 1rem;
          margin-bottom: 1.5rem;
        `}
        gap="1rem"
        alignItems="center"
      >
        <div css={css`width: 250px`}>
          <Input type="text" name="search" placeholder="Search..." onChange={onSearchChange} />
        </div>
        <div css={css`margin-left: 1rem;`}>
          <Checkbox
            checked={state.settings.showClosedIssues}
            onChange={toggleClosedIssues}
            label="Include closed"
          />
        </div>
      </Flex>
      <IssuesTable search={search} />
    </Flex>
  );
};

export { SummaryPage };
