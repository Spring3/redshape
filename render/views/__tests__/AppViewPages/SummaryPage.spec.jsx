import React from 'react';
import { HashRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import configureStore from 'redux-mock-store';
import MockAdapter from 'axios-mock-adapter';
import { ThemeProvider } from 'styled-components';

import theme from '../../../theme';
import SummaryPage from '../../AppViewPages/SummaryPage';
import issueActions, { ISSUES_GET_PAGE } from '../../../actions/issues.actions';
import { initialize, getInstance, reset } from '../../../../common/request';

const mockStore = configureStore([thunk]);
const redmineEndpoint = 'redmine.test.test';
const token = 'multipass';

let axiosMock;

describe('AppView -> Summary Page', () => {
  beforeAll(() => {
    initialize(redmineEndpoint, token);
    axiosMock = new MockAdapter(getInstance());
  });

  afterEach(() => {
    axiosMock.reset();
    cleanup();
  });

  afterAll(() => {
    axiosMock.restore();
    reset();
  });

  it('should fetch the issues on mount', () => {
    axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, {}]));
    const spy = jest.spyOn(issueActions, 'getPage');
    const state = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Subject', isFixed: true, value: 'subject' }
        ]
      }
    };
    const store = mockStore(state);
    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <HashRouter>
            <SummaryPage />
          </HashRouter>
        </ThemeProvider>
      </Provider>
    );

    expect(spy).toHaveBeenCalled();
    expect(store.getActions()[0].type).toBe(ISSUES_GET_PAGE);
    expect(store.getActions()[0].info.page).toBe(0);
    spy.mockRestore();
  });

  it('should fetch issues on search', (done) => {
    axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, {}]));
    const state = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Subject', isFixed: true, value: 'subject' }
        ]
      }
    };
    const store = mockStore(state);
    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <HashRouter>
            <SummaryPage />
          </HashRouter>
        </ThemeProvider>
      </Provider>
    );

    fireEvent.change(document.querySelector('input[name="search"]', {
      target: {
        value: 'search'
      }
    }));

    setTimeout(() => {
      expect(store.getActions().length).toBe(2);
      expect(store.getActions().filter((action) => action.type === ISSUES_GET_PAGE).length).toBe(2);
      done();
    }, 1);
  });

  it('should set the sorting properties', (done) => {
    axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, {}]));
    const state = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Subject', isFixed: true, value: 'subject' }
        ]
      }
    };
    const store = mockStore(state);
    const { getAllByText } = render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <HashRouter>
            <SummaryPage />
          </HashRouter>
        </ThemeProvider>
      </Provider>
    );

    fireEvent.click(getAllByText('Id').pop());
    setTimeout(() => {
      expect(store.getActions().length).toBe(2);
      expect(store.getActions().filter((action) => action.type === ISSUES_GET_PAGE).length).toBe(2);
      done();
    }, 1);
  });

  it('should fetch the issues when showClosedIssues state changes', (done) => {
    axiosMock.onGet('/issues.json').reply(() => Promise.resolve([200, {}]));
    const state = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        redmineEndpoint: 'https://redmine.domain',
        api_key: '123abc'
      },
      issues: {
        all: {
          data: []
        }
      },
      settings: {
        issueHeaders: [
          { label: 'Id', isFixed: true, value: 'id' },
          { label: 'Subject', isFixed: true, value: 'subject' }
        ]
      }
    };
    const store = mockStore(state);
    render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <HashRouter>
            <SummaryPage />
          </HashRouter>
        </ThemeProvider>
      </Provider>
    );

    fireEvent.click(document.querySelector('#queryOptions input:first-child'));
    setTimeout(() => {
      expect(store.getActions().length).toBe(3);
      expect(store.getActions().filter((action) => action.type === ISSUES_GET_PAGE).length).toBe(2);
      done();
    }, 1);
  });
});
