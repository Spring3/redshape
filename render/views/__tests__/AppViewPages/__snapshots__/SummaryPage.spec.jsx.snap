// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppView -> Summary Page should match the snapshot 1`] = `
.c3 {
  display: block;
  width: 100%;
  border-radius: 3px;
  padding: 5px 10px;
  box-sizing: border-box;
  font-size: 14px;
  min-height: 35px;
  outline: none;
  font-weight: bold;
  -webkit-transition: background .2s;
  transition: background .2s;
  border: 1px solid #A0A0A0;
  color: #FF7079;
  background: white;
  background-image: url(data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20class%3D%22mdi-icon%20%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23FF7079%22%20viewBox%3D%220%200%2024%2024%22%3E%3Cpath%20d%3D%22M9.5%2C3A6.5%2C6.5%200%200%2C1%2016%2C9.5C16%2C11.11%2015.41%2C12.59%2014.44%2C13.73L14.71%2C14H15.5L20.5%2C19L19%2C20.5L14%2C15.5V14.71L13.73%2C14.44C12.59%2C15.41%2011.11%2C16%209.5%2C16A6.5%2C6.5%200%200%2C1%203%2C9.5A6.5%2C6.5%200%200%2C1%209.5%2C3M9.5%2C5C7%2C5%205%2C7%205%2C9.5C5%2C12%207%2C14%209.5%2C14C12%2C14%2014%2C12%2014%2C9.5C14%2C7%2012%2C5%209.5%2C5Z%22%3E%3C%2Fpath%3E%3C%2Fsvg%3E);
  background-repeat: no-repeat;
  background-position: 5px center;
  padding-left: 35px;
}

.c3:hover {
  border: 1px solid #FF7079;
}

.c3:focus {
  border-color: #FF7079;
  box-shadow: 0px 0px 0px 1px #FF7079;
}

.c3::-webkit-input-placeholder {
  color: #A0A0A0;
  font-weight: 500;
}

.c3::-moz-placeholder {
  color: #A0A0A0;
  font-weight: 500;
}

.c3:-ms-input-placeholder {
  color: #A0A0A0;
  font-weight: 500;
}

.c3::placeholder {
  color: #A0A0A0;
  font-weight: 500;
}

.c7 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clippath: inset(50%);
  clippath: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c8 {
  display: inline-block;
  width: 16px;
  height: 16px;
  background: #F9F9F9 transition:background .2s;
  border: 2px solid transparent;
  border-radius: 3px;
  cursor: pointer;
  background: white;
  border-color: #FF7079;
}

.c8 svg {
  visibility: hidden;
  position: relative;
  vertical-align: middle;
  bottom: 3px;
}

.c6:not(:disabled):hover + .c8,
.c6:focus + .c8 {
  box-shadow: 0px 0px 5px 1px #FF7F87;
}

.c6:disabled + .c8 {
  background: #F9F9F9;
  border-color: lightgrey;
}

.c5 {
  display: inline-block;
  vertical-align: middle;
}

.c5 + span {
  margin-left: 10px;
  vertical-align: middle;
}

.c9 {
  position: relative;
  min-height: 200px;
  width: 100%;
  border: 2px solid #F9F9F9;
  border-spacing: 0px;
}

.c9 tbody {
  text-align: center;
  font-size: 14px;
  overflow-y: scroll;
  max-height: 60vh;
}

.c9 tbody tr td {
  border-bottom: 2px solid #F9F9F9;
  padding: 15px 5px;
  -webkit-transition: background ease .2s;
  transition: background ease .2s;
}

.c9 tbody tr:hover {
  cursor: pointer;
}

.c9 tbody tr:hover td {
  background: #F9F9F9;
}

.c9 tbody tr td.due_date {
  white-space: nowrap;
}

.c9 tbody tr td.subject {
  text-align: left;
}

.c9 thead {
  z-index: 1;
}

.c9 thead tr th {
  padding: 15px 5px;
  background: #F9F9F9;
  border-bottom: 2px solid #E9E9E9;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
}

.c9 thead tr th:hover {
  cursor: pointer;
  color: #FF7079;
}

.c9 thead tr th svg {
  vertical-align: middle;
}

.c9 thead tr th.subject {
  text-align: left;
}

.c0 {
  display: grid;
  padding: 20px;
  grid-template-columns: repeat(auto-fit,minmax(100px,1fr));
  margin-bottom: 60px;
}

.c1 {
  background: white;
  padding: 0px 20px 20px 20px;
  border-radius: 5px;
  grid-column: span 8;
  grid-row: auto;
}

.c2 {
  display: grid;
  grid-template-columns: 1fr minmax(80px,auto);
  grid-template-rows: auto;
  grid-row-gap: 20px;
  grid-column-gap: 20px;
  margin-bottom: 20px;
}

.c4 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

<div
  className="c0"
>
  <section
    className="c1"
  >
    <h2>
      Issues assigned to me
    </h2>
    <div
      className="c2"
    >
      <input
        className="c3"
        disabled={false}
        name="search"
        onChange={[Function]}
        placeholder="Search..."
        type="text"
      />
      <div
        className="c4"
      >
        <label>
          <div
            className="c5"
          >
            <input
              checked={false}
              className="c6 c7"
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
            <div
              checked={false}
              className="c8"
            >
              <svg
                className="mdi-icon "
                fill="white"
                height="18"
                viewBox="0 0 24 24"
                width="18"
              >
                <path
                  d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
                />
              </svg>
            </div>
          </div>
          <span>
            Closed
          </span>
        </label>
      </div>
    </div>
    <table
      className="c9"
    >
      <thead>
        <tr>
          <th
            className=""
            onClick={[Function]}
          >
            Id
             
          </th>
          <th
            className=""
            onClick={[Function]}
          >
            Subject
             
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </section>
</div>
`;
